README.md
# GRIPT - Angular Gulp Sass Typescript application builder

[![npm](https://img.shields.io/npm/v/gript.svg)](http://npmjs.com/package/gript)
[![npm](https://img.shields.io/npm/dm/gript.svg)](http://npmjs.com/package/gript)
[![npm](https://img.shields.io/npm/dt/gript.svg)](http://npmjs.com/package/gript)

<img align="right" src="package/img/gript_logo_big.png">

This project is the basis for the creation of an [npm module](https://www.npmjs.com/package/gript) containing a "complete" [Gulp](http://gulpjs.com) build setup for
an [AngularJS](https://angularjs.org) application using [Sass](http://sass-lang.com). Your application source and test files can be written in JavaScript or [TypeScript](http://www.typescriptlang.org/).
If it's TypeScript, it will be validated and then compiled to JavaScript.

The goal of this tool is to provide the handy workflow for developing Angular based applications (no matter what language you use - JavaScript or TypeScript)
and ensure the profound checking of the code quality at the same time.
The module orchestrates a collection of [Gulp](http://gulpjs.com) build functionality into a single npm dependency
and makes it easier for developers to have and maintain the build setup for own Angular projects.

The [Gript Newsgroup](https://groups.google.com/forum/#!forum/gript) is the place for announcements and discussion about Gript's features. Feel free to join.
 
The project includes a sample application to make it possible to test the build setup.
The local sample application exemplifies the needed structure for applications supported by the `gript` npm package.
The sample application resides in the `app` folder. 
The `index.html` in the `app` folder is especially important - it contains markers where generated files will be included.

## Features

- JavaScript validation using [ESLint](http://eslint.org),
- TypeScript validation ([TSLint](http://palantir.github.io/tslint/)) and incremental compilation,
- scss validation ([scss-lint](https://github.com/brigade/scss-lint)) and compilation,
- HTML validation using [htmllint](https://github.com/htmllint/),
- unit tests performed using [Karma](https://karma-runner.github.io) and [PhantomJS](http://phantomjs.org) (tests can be written in JavaScript or TypeScript),
- unit testing coverage metered by [Istanbul](https://github.com/gotwarlost/istanbul),
- HTML partials pre-loading into the Angular `$templateCache`,
- includes custom generated [Modernizr](https://modernizr.com) script
- full concatenation/minification for all production JS and CSS files,
- Live-reload capability: web app is auto-refreshed if HTML, TypeScript or scss files change,
- watch tasks: if your source files change, they will be checked for errors, compiled and then injected into your application,
 Gript uses [Chokidar](https://github.com/paulmillr/chokidar) which notices the changes instantly, keeping the CPU usage down at the same time,
- contains the mock server based on `YAML/JSON/JS` configuration files. Refer to the [Mock Server](#mocks) section for usage guidelines.
- contains the locked set of specific `npm` dependencies, to minimize "the dependency hell".

## Structure requirements for applications supported

The use of the this Gulp build tool is based on applications code being structured according to Google's
 [Best Practice Recommendations for Angular App Structure](https://docs.google.com/document/d/1XXMvReO8-Awi1EZXAXS4PzDzdNvV6pGcuaF4Q9821Es/mobilebasic?pli=1). The primary directories and files are:

    |---- /app
    |     |
    |     |---- components
    |     |---- sections
    |     |---- img
    |     |---- styles
    |     |---- resources
    |     |---- lib
    |     |---- index.html
    |     |---- app.js
    |     |---- .eslint.rc.yml
    |     |---- constants.json
    |---- /bower_components
    |---- /target
    |     |---- dist
    |     |---- tmp
    |---- /mocks
    |---- gulpfile.js
    |---- bower.json
    |---- package.json
    |---- .eslint.rc.yml
    |---- .scsslint.rc.yml
    |---- tslint.json
    |---- .htmllint.rc

which means:

- `app` : contains the application source code
- `app/sections` : contains the subsections of the application code
- `app/components`: contains the components (directives, services etc.) embedded in the application
- `app/app.js` or `app/App.ts` : the entry point of the [Angular](https://angularjs.org) application
- `app/resources`: the place for other resources, like translation files. This will be copied to /target/dist
- `app/lib`: the place for JavaScript libraries not coming from Bower. It will be excluded from linting.
- `app/constants.json`: optional constants file from which Angular constants module will be generated
- `bower_components` : libraries downloaded by [Bower](http://bower.io/)
- `node_modules` : tools downloaded by [npm](https://www.npmjs.org/)
- `target/tmp` : contains generated files (compiled TypeScript, compiled scss styles, Angular templates etc.)
- `target/dist` : contains app distribution package
- `mocks` : contains your mock services definitions.
- `gulpfile.js` : the build files importing the [Gulp](http://gulpjs.com) tasks defined in the `node_modules/gript`
- `bower.json` : contains [Bower](http://bower.io) dependencies
- `.eslint.rc.yml` : contains rules for es-linter ([ESLint](http://eslint.org)). Cascading rules configuration is possible.
- `.scsslint.rc.yml` : contains rules for scss linter ([scss-lint](https://github.com/brigade/scss-lint))
- `tslint.json` : rules for the TypeScript linter ([TSLint](http://palantir.github.io/tslint/))
- `.htmllint.rc` : rules for the HTML linter ([htmllint](https://github.com/htmllint))

## Setup GRIPT in your project

To make use of the `gript` npm module define the dependency to `gript` in the `package.json` file in your own application:

      "dependencies": {
          "gulp": "3.9.0",
          "gript": "~0.0.38",
          ....

together with your other npm dependencies. Gript is available in the [npm repository](https://www.npmjs.com/package/gript).

Running the command

    `npm install --save-dev`

will include `gript` in your `node_modules` folder and make it available for development.

## Configure your project

The tool comes with a guide for the enabling of the build tool in your project.

#### The sample Gulp file and configuration files

The supported Angular application is built by [Gulp](http://gulpjs.com), which is controlled by the `gulpfile.js`.
Gript includes a sample Gulp file, located in the `sample_configs` directory.
There are also example configuration files for `Bower`, `es-lint`, and `scss-lint` in the `sample_configs` directory.

#### Using the sample Gulp file
This `sample_configs/gulpfile.js` can be used as a starter for your project. This is where you define the dependency to `gript` module and specify the options for the build process of your own application:

```javascript
       var gulp = require('gulp');
       
       require('gript')(gulp);
       
       // Set the config to use across the gulp build
       gulp.config = {
           repository: 'http://git.nykreditnet.net/scm/dist/xpa-no-specified-project.git',
           app: {
               module: 'yourApp',
               constantsFile: 'app/constants.json'
           },
           server: {
			   module: 'yourApp',
			   configFile: 'app/config.json'
		   },
		   partials: ['app/**/*.html'],
		   server: {
               port: 8080,
               host: 'localhost',
               livereload: {
                   port: 35729
               }
           },
           serverDist: {
               port: 8080,
               host: 'localhost'
           },
           proxy: {
               port: 8001,
               hostHeader: 'gript',
               targetURL: 'http://gript'
           },
           mocks: {
               stubby: {
                   location: 'localhost',
                   stubs: 8050,
                   tls: 8443,
                   admin: 8051,
                   relativeFilesPath: true,
                   files: [
                       'mocks/*.{json,yaml,js}'
                   ]
               },
               multimocks: {
                   src: 'multiMocks'
               }
           },
           typeScript: {
               compilerOptions: {
                   noImplicitAny: true,
                   target: "es5",
                   sourceMap: true,
                   declarationFiles: true,
                   noExternalResolve: false,
                   sortOutput: true,
                   removeComments: false,
                   preserveConstEnums: true
               }
           },
           minification: {
               html: {
                   removeEmptyAttributes: false,
                   collapseBooleanAttributes: false,
                   collapseWhitespace: true,
                   caseSensitive: true
               },
               css: {
                   safe: true,
                   autoprefixer: false,
                   discardUnused: false,
                   reduceIdents: false,
                   mergeIdents: false
               },
               javascript: {
                   mangle: true,
                   preserveComments: false
               }
           },
           modernizr: {
               //cssprefix: true, // add this line to fix conflicts with bootstrap (e.g not adding class 'hidden' to html tag)
               options: [
                   'addTest',
                   'html5printshiv',
                   'testProp',
                   'fnBind',
                   'setClasses'
               ],
               'feature-detects': []
		   },
           fontsScan: [
               'bower_components/font-awesome', 
               'bower_components/bootstrap-sass-official'
           ]
           karma: {
               browsers: ['PhantomJS']
           }
	   };
```

Be sure to set values for the configuration in your copy of the `sample_configs/gulpfile.js`.

These values are:

- `app.module` : mandatory name of the project. It's being used as a module name when generating Angular modules, like `$templateCache` or constants modules.
- `repository` the GIT url of your application, used in the `release` and `prerelease` tasks.
- `partials` is a glob pattern to specify what files should bne considered as Angular `$templateCache` templates. Refer to the [Partials](#partials) section for details.
- `server` configuration options for the web server like port number, live reload port number, host name etc.
- `serverDist` configuration options for the web server started from `dist` by using `server:dist` task.
- `proxy` optional proxy port, host header and target URL configuration. The proxy will start only if you specify `proxy` section in your `gulpfile.js`.
- `mocks` mock server configuration
- `typescript` typescript compilation options
- `minification` minification related options
 
You may kickstart your project by copying `sample_configs/gulpfile.js` to the root of your own project.
This gives you a very simple build configuration as a starting scenario.

NOTE: If you have no tests the

    gulp test

command will fail, until you create your first piece of logic and its corresponding test.

## Building your project

The **default** task builds an application and then starts the development server. Your source code (TypeScript, HTML, scss) will be watched for changes, and - if neccessary, compiled and injected into the `index.html`.
Just by running

    gulp

you are ready to start developing your project.

The `gulpfile.js` from Gript contains also these specific tasks:

- **build** : builds the application for the development
- **dist** : builds and minifies the application for the deployment. The application will be copied to `target/dist` directory.
- **ts** : compiles your app TypeScript files
- **partials** : compiles HTML partials into Angular's `$templateCache` Javascript files. All files matching `partials` config glob are considered as templates.
- **styles** : compiles scss files
- **inject** : injects Bower dependencies, compiled HTML partials, TypeScript and scss into your app's `index.html`. Files will be injected according to the marking in the `index.html` file. Refer to the [Files injection](#injection) section of this readme for details.
    - **inject-bower** : downloads and injects [Bower](http://bower.io/) dependencies
    - **inject-styles** : compiles and injects scss styles
    - **inject-partials** : compiles HTML partials into Angular's `$templateCache` and then injects them into the `index.html`
    - **inject-js** : complies the TypesScript and then injects all JavaScript files
- **lint** : runs linters on the HTML, scss, Javascript and TypeScript source files. Refer to the [Linting](#linting) section for possible options.
	- **lint-js**
	- **lint-scss**
	- **lint-ts**
	- **lint-html**
- **test** : runs the unit tests through [Karma](http://karma-runner.github.io) - NOTE: fails if no tests are available. Refer to the [Testing](#testing) section for details.
- **clean** : removes the whole `target` directory (temporary generated files and distribution package)
    - **clean-dist** : removes the `target/dist` directory (the distribution package)
    - **clean-tmp** : removes the `target/tmp` directory (all temporary generated files)
    - **clean-js** : removes the `target/tmp/js` directory (compiled TypeScript files)
    - **clean-partials** : removes the `target/tmp/partials` directory (Angular's `$templateCache` Javascript files)
    - **clean-styles** : removes the `target/tmp/styles` directory (compiled scss files)
    - **clean-bower** : removes the `bower_components` directory
- **fonts** : searches for all `eot`, `ttf`, `woff` , `woff2` files, flattens the directory structure and copies them into your app. It will search the whole `bower_components` directory, unless you configure it otherwise in the `fontsScan` array in the config section of the `gulpfile.js`.
- **images** : copies all image files into the `dist` directory
- **watch** : watches the source code for changes and runs the relevant task(s) whenever something changes
- **server** : starts a development server
- **server:dist** : starts a server using the deployment directory (`target/dist`)
- **mocks** : starts a server with mock services. Refer to the [Mock server](#mocks) section for guidelines.
- **modernizr** : builds custom [Modernizr](http://modernizr.com) script and injects it into `index.html`
- **config** : creates an optional Angular constants module with values from the file specified in `gulp.config.app.constantsFile`. The default is `app/constants.json`.

You can list all of the available tasks by running the command:

    gulp --tasks

<a name="testing"></a>
## Testing
Gript will automatically run your tests using [Karma](https://karma-runner.github.io). Your tests can be written in JavaScript or TypeScript (they will be compiled first). Tests filenames must end in `test` or `Test` (for example `PortfolioServiceTest.ts`, `PortfolioService_test.ts`, `portfolioService_test.js`).
Your tests will be executed using headless, [PhantomJS](http://phantomjs.org) browser. You can customize this behavior altogether with other Karma options in the `karma` section of the `gulpfile.js`.
Refer to [Karma Configuration docs](https://karma-runner.github.io/0.13/config/configuration-file.html) for possible options and values. 
For example, if you need to run your tests under Chrome browser, amend the default configuration:
 
 ```
  karma: {
         frameworks: ['jasmine', 'sinon', 'angular-filesort'],
         plugins: [
             'karma-jasmine',
             'karma-sinon',
             'karma-angular-filesort',
             'karma-phantomjs-launcher',
             'karma-chrome-launcher',
             'karma-junit-reporter',
             'karma-coverage',
             'karma-ng-html2js-preprocessor'
         ],
         browsers: ['Chrome']
     }
 ```

<a name="partials"></a>
## Partials
The `partials` task will create Angular `$templateCache` files from your HTML files. The resulting JavaScript files will be created in the `target/tmp/partials` directory.
These files will then be injected into the `index.html` file, according to `<!-- inject:partials --><!-- endinject -->` markings.
The `dist` task will minify and concatenate them with other JavaScript files from your application.
Gript considers all HTML files in the `app` directory (except the `index.html`) as partials by default. You can change this behaviour by setting up the `partials` configuration value in the `gulpfile.js`. This may come in handy when you don't want a specific HTML file to be converted to Angular `$templateCache` partial.
For example:
```
partials: ['app/**/*.html', '!app/sections/welcome/testOauth.html']
```
 
will generate `$templateCache` from all HTML files except the `app/sections/welcome/testOauth.html`.
Take note that all files excluded from partials generation will be just copied to the `target` directory. At the same time, all files considered as partials will not be copied - after the conversion to JavaScript we don't need them anymore.

If you need to customize the partials generation, use the `partialsOptions` in the `gulpfile.js`, for example:

```
partials: ['app/**/*.html'],
partialsOptions: {
    prefix: './'
},
```

Refer to the [gulp-ng-html2js](https://www.npmjs.com/package/gulp-ng-html2js) documentation for possible options. 

<a name="linting"></a>
## Linting
To ensure the profound checking of the code quality of your application, Gript will check all your HTML, scss, TypeScript and JavaScript files.
The linting process is executed during the build, and is also included in the `watch` task, to re-lint the file on the fly, after you change it.
The number of configuration files are being used to customize the linting options:

- `.eslint.yml` contains configuration for the powerful JavaScript linter, the [ESLint](http://eslint.org). Refer to the [Options](http://eslint.org/docs/user-guide/configuring) section for avaialable options.
- `.scss-lint.yml` contains configuration for scss linter, the [scss-lint](https://github.com/brigade/scss-lint). Referer to the [Configuration](https://github.com/brigade/scss-lint#configuration) for options.
- `tslint.json` contains options for [TSLint](http://palantir.github.io/tslint/). Refer [here](http://palantir.github.io/tslint/rules/) for the description of rules.
- `.htmllintrc` contains setup for [htmllint](https://github.com/htmllint). Refer to the [options](https://github.com/htmllint/htmllint/wiki/Options) for possible settings.

For your convenience, Gript contains sample configuration files for all linters in the `sample_configs` directory.
During the linting process, Gript generates Checkstyle-like XML reports in the `target` directory:

- `es-lint-result.xml`
- `html-lint-result.xml`
- `scss-lint-result.xml`
- `ts-lint-result.xml`

The reports from tests execution and code coverage are also generated into the `target` directory:

- `cobertura-coverage.xml`
- `test-results.xml`

You can use linting, test and coverage reports in your continuous integration setup, like [Jenkins](https://jenkins-ci.org) for example.

## Modernizr
During the build process, Gript will analyse your JavaScript and CSS files and generate custom [Modernizr](http://modernizr.com).
The custom script will be injected into the `index.html`.
You can customize the script generation by changing options in the `modernizr` section of the `gulpfile.js`.
Refer to the [Modernizr build options](https://github.com/Modernizr/Modernizr/blob/master/lib/config-all.json) for possible values.

## TypeScript compilation
If you develop your app in the TypeScript, files will be compiled and then injected. The example setup uses [DefinitelyTyped](http://definitelytyped.org/)
 to get the TypeScript types definitions. They are being downloaded by [Bower](http://bower.io/) into the `bower_components/DefinitelyTyped` directory, which is excluded from the TypeScript linting.
 The resulting JavaScript files will be placed in the `target/tmp/js` directory.
 You can customize your TypeScript compile options using the `typeScript` section in the `gulpfile.js`.
 Refer to the [Compiler-Options](https://github.com/Microsoft/TypeScript/wiki/Compiler-Options) section in the TypeScript documentation for available options.

<a name="minification"></a>
## Minification
After executing a `dist` task, all resulting HTML, CSS and JavaScript files will be concatenated and minified.
You can customize the minification options by modyfying the `minification` section in your `gulpfile.js`.

- HTML files will be minified using [html-minifer](https://github.com/kangax/html-minifier). See the [html-minifier docs](https://github.com/kangax/html-minifier) for available options.
- CSS files will be minified with [cssnano](https://github.com/ben-eb/cssnano). Refer to the [documentation](http://cssnano.co/optimisations/) for possible optimisations.
- JavaScript will be compressed with [UglifyJS](http://lisperator.net/uglifyjs/). Refere to the [gulp-uglify](https://www.npmjs.com/package/gulp-uglify) module docs for help how to set options.

:warning: Note: too agressive optimisations will break your build or make your app useless. Gript includes the default set of options, which are pretty safe.

<a name="injection"></a>
## Compiled files injection

By default, all TypeScript, scss and HTML files will be compiled and injected into your app's `index.html` file.
The `target/tmp` directory is the place for the compilation output:

- `target/tmp/js` : will contain compiled TypeScript files
- `target/tmp/partials` : will contain all HTML partials from your Angular app compiled into Angular `$templateCache`.
- `target/tmp/styles` : will contain compiled scss files

Compiled scripts and styles will then be injected into the app's `index.html` according to the injection markings:

- `<!-- bower:css --><!-- endbower -->` : Bower CSS dependencies
- `<!-- inject:css --><!-- endinject -->` : compiled scss files
- `<!-- bower:js --><!-- endbower -->` : Bower JS dependencies
- `<!-- inject:js --><!-- endinject -->` : TypeScript files compiled into JS
- `<!-- inject:partials --><!-- endinject -->` : HTML partials compiled into Angular's `$templateCache`.

After executing the `dist` task, all the stylesheets and JavaScripts will be minimized and concatenated.
The result will be injected into `target/dist/index.html` according to these injection markings:
- `<!-- build:css styles/vendor.css --><!-- endbuild -->`: vendor stylesheets (from `bower_components`)
- `<!-- build:css styles/main.css --><!-- endbuild -->`: your own stylesheets
- `<!-- build:js scripts/vendor.js --><!-- endbuild -->`: vendor scripts (from `bower_components`)
- `<!-- build:js scripts/main.js --><!-- endbuild -->`: your own scripts

Refer to `app/index.html` or `sample_configs/index.html` for an example how to define these markings.

<a name="mocks"></a>
## Mock server
There are two mock engines implementations included in Gript: [stubby4node](https://github.com/mrak/stubby4node) and [Angular-Multimocks](https://github.com/wongatech/angular-multimocks).
[stubby4node](https://github.com/mrak/stubby4node) allows you to have a mock server running during the development, using the standard `gulp` or `gulp server` tasks.

CHANGES.md
## Gript 0.0.53 (27-06-2016)

- Enabled custom params when starting protractor.
- fixed [#70](https://github.com/Nykredit/gript/issues/70). modernizr.js, bower_components and node_modules excluded from coverage.
 

## Gript 0.0.52 (13-05-2016)

- implemented [#77](https://github.com/Nykredit/gript/issues/77).  `gulp-ng-html2js` options are now exposed in the config in the `gulpfile.js`. 

## Gript 0.0.51 (12-05-2016)

- fixed [#76](https://github.com/Nykredit/gript/issues/76) Gript proxy crashes

## Gript 0.0.50 (09-05-2016)

- implemented [#68](https://github.com/Nykredit/gript/issues/68) "Making an api mock without starting a real server"
There are two mock engines implementations included in Gript now: runtime [stubby4node](https://github.com/mrak/stubby4node) and generation-based [Angular-Multimocks](https://github.com/wongatech/angular-multimocks).
You have an option to start both of them, just a single one or none.

## Gript 0.0.49 (27-04-2016)

- proxy task is now optional. It will start only if you specify `proxy` section in your `gulpfile.js`. 

## Gript 0.0.48 (13-04-2016)

- implemented [#67](https://github.com/Nykredit/gript/issues/67). Karma settings are now exposed in the `gulpfile.js`, so you can run your tests in Chrome, for example. The `tasks/test.gulp.js` contains the default Karma configuration. 

## Gript 0.0.47 (08-04-2016)

 - dependencies updated:
    - phantomjs: 2.1,
    - karma": 0.13.22,
    - karma-coverage: 0.5.5,
    - karma-jasmine: 0.3.8,
    - karma-spec-reporter: 0.0.26

## Gript 0.0.46 (08-04-2016)

 - fixed [#66](https://github.com/Nykredit/gript/issues/66) - missing Karma plugins in the `package/karma.conf.js` 
 - implemented [#42](https://github.com/Nykredit/gript/issues/42). Sinon is now loaded as Karma plugin in `karma.conf.js`. Reference to aboslute path to `sinon.js` in the `test` task removed.
 
## Gript 0.0.45 (07-04-2016)
fixed a `dist` path typo

## Gript 0.0.44 (06-04-2016)
`$templateCache` partials are now configurable.
Gript considers all HTML files in the `app` directory (except the `index.html`) as partials by default. 
You can change this behaviour by setting up the `partials` configuration value in the `gulpfile.js`.
This may come in handy when you don't want a specific HTML file to be converted to Angular `$templateCache` partial.


## Gript 0.0.43 (31-03-2016)

 - implemented [#60](https://github.com/Nykredit/gript/issues/60). You can prevent the mocks server form starting by using the `--nomocks` command line option.

## Gript 0.0.42 (30-03-2016)

 - implemented [#46](https://github.com/Nykredit/gript/issues/46). From now on, Gript will continue with the default rules and issue a warning if any of the linter configuration is missing:
    - scss-lint.yml
    - .htmllintrc
    - tslint.json
 - merged [#40](https://github.com/Nykredit/gript/pull/40)
 
## Gript 0.0.41 (25-03-2016)

 - implemented **#6** (Automate release process). Gript's own `gulpfile.js` contains new tasks:
 	- `bumpVersion` - bumps package version number in `package.json` and `package/package.json`
 	- `tag` - Tags the git repository with the version number from `package.json`
 	- `pushTags` - pushes tags to `master`
 	- `push` - pushes source to `master`
 	- `npmPublish` - publishes to npm.js. This is an interactive task, asks for npm username and password
 	- `publish` - executes all of the above in the sequence.
 - `Gulp` upgraded to 3.91.
 
## Gript 0.0.40 (25-03-2016)

 - fixed **#58** (html-lint doesn't pick up all files)

## Gript 0.0.39 (24-03-2016)

 - proxy port can now be customized in the config.

## Gript 0.0.38 (23-03-2016)

 - dist server port and hostname can now be customized in the gulpfile.js config
 - file created by `config` task renamed from `config.js` to `constants.js`. The task searches for `constants.json` file by default.
 - starting with this release, only `*.tpl.html` files will be considered as Angular templates and converted to Angular's cache service javascripts. All other HTML files will be just copied into the dist structure. The reason for that is that some apps need HTML files which are not Angular templates.


gulpfile.js
'use strict';

var bump = require('gulp-bump'),
    gulp = require('gulp'),
    git = require('gulp-git'),
    ex = require('gulp-expect-file'),
    merge = require('merge-stream'),
    path = require('path'),
    sequence = require('run-sequence').use(gulp),
    spawn = require('child_process').spawn,
    tagVersion = require('gulp-tag-version'),
    projectRoot = process.cwd(),
    paths = {
        bower: 'bower_components',
        src: {
            app: 'app',
            index: 'app/index.html',
            scss: 'app/app.scss',
            constants: 'app/constants.json',
            img: 'app/**/img/**/*',
            resources: 'app/resources/**/*',
            lib: 'app/lib/**/*',
            fonts: '**/*.{eot,ttf,woff,woff2}',
            mocks: 'mocks/*.{json,yaml,js}'
        },
        target: {
            base: 'target',
            tmp: {
                base: 'target/tmp',
                js: 'target/tmp/js',
                partials: 'target/tmp/partials',
                styles: 'target/tmp/styles',
                resources: 'target/tmp/resources',
                lib: 'target/tmp/lib',
                fonts: 'target/tmp/fonts'
            },
            dist: {
                base: 'target/dist',
                resources: 'target/dist/resources',
                lib: 'target/dist/lib',
                fonts: 'target/dist/fonts'
            }
        },
        javaScriptToInject: [
            'app/**/*.js', //javascript source
            'target/tmp/js/**/*.js', //compiled TypeScript
            '!target/tmp/js/**/*Test.js',
            '!target/tmp/js/**/*test.js',
            '!app/**/*Test.js',
            '!app/**/*test.js'
        ],
        linters: {
            scss: '.scss-lint.yml',
            eslint: '.eslintrc.yml',
            html: '.htmllintrc',
            ts: 'tslint.json'
        },
        protractor: {
            config: 'e2e-tests/protractor-config.js',
            tests: './e2e-tests/**/*_e2e-test.js'
        }
    };

require("fs").readdirSync(__dirname + "/tasks").forEach(function (file) {
    if (file.indexOf("gulp") > -1) {
        require(__dirname + "/tasks" + "/" + file)(gulp, paths);
    }
});

// Register the default task
gulp.task('default', function () {
    sequence('build', ['server', 'proxy'], 'watch');
});

gulp.task('dist:serve', ['dist', 'server:dist']);


// Set the config to use across the gulp build
gulp.config = {
    repository: 'http://nykredit.github.com/gript.git',
    app: {
        module: 'gript'
    },
    partials: ['app/**/*.html'],
    server: {
        port: 8080,
        host: 'localhost',
        livereload: {
            port: 35729
        }
    },
    serverDist: {
        port: 8080,
        host: 'localhost'
    },
    proxy: {
        port: 8001,
        targetURL: 'http://gript',
        hostHeader: 'gript'
    },
    typeScript: {
        compilerOptions: {
            noImplicitAny: true,
            target: "es5",
            sourceMap: true,
            declarationFiles: true,
            noExternalResolve: false,
            sortOutput: true,
            removeComments: false,
            preserveConstEnums: true
        }
    },
    minification: {
        html: {
            removeEmptyAttributes: false,
            collapseBooleanAttributes: false,
            collapseWhitespace: true,
            caseSensitive: true
        },
        css: {
            safe: true,
            autoprefixer: false,
            discardUnused: false,
            reduceIdents: false,
            mergeIdents: false
        },
        javascript: {
            mangle: true,
            preserveComments: false
        }
    },
    fontsScan: [
        'bower_components/font-awesome',
        'bower_components/bootstrap-sass-official'
    ]
};

/**
 * Bumps package version number in package.json and package/package.json
 */
gulp.task('bumpVersion', function () {
    var stream1 = gulp.src(path.join(projectRoot, 'package.json'))
            .pipe(bump())
            .pipe(gulp.dest(path.join(projectRoot))),

        stream2 = gulp.src(path.join(projectRoot, 'package', 'package.json'))
            .pipe(bump())
            .pipe(gulp.dest(path.join(projectRoot, 'package')));

    return merge(stream1, stream2);
});

/**
 * Tags the git repository with the version number from package.json
 */
gulp.task('tag', function () {
    git.commit('bumps package version', {cwd: projectRoot});

    return gulp.src(path.join(projectRoot, 'package.json'))
    .pipe(tagVersion({cwd: projectRoot}));
});

function gitPush(args) {
    return git.push('origin', 'master', {args: args}, function (err) {
        if (err) {
            throw err;
        }
    });
}

gulp.task('pushTags', function () {
    return gitPush(' --tags');
});

gulp.task('push', function () {
    return gitPush(' -f');
});

/**
 * Publishes to npm repository. This is an interactive task, asks for npm username and password.
 */
gulp.task('npmPublish', function (done) {
    spawn('npm', ['publish', path.join(projectRoot, 'target', 'gript')], { stdio: 'inherit' }).on('close', done);
});
/**
 * Increases version number, tags the repository, pushes source and tags to master and then publishes to npm.js
 */
gulp.task('publish', function () {
    return sequence('bumpVersion', 'tag', 'push', 'pushTags', 'npmPublish');
});

// verify that the build setup can produce the expected artifacts
gulp.task('verify-package-foundation', ['build'], function () {
    gulp.src('target/dist/index.html')
        .pipe(ex.real('target/dist/index.html'));
    gulp.src('target/dist/app.manifest')
        .pipe(ex.real('target/dist/app.manifest'));

    gulp.src('target/dist/styles/main-*.css')
        .pipe(ex.real('target/dist/styles/main-*.css'));
    gulp.src('target/dist/styles/vendor-*.css')
        .pipe(ex.real('target/dist/styles/vendor-*.css'));

    gulp.src('target/dist/scripts/main-*.js')
        .pipe(ex.real('target/dist/scripts/main-*.js'));
    gulp.src('target/dist/scripts/vendor-*.js')
        .pipe(ex.real('target/dist/scripts/vendor-*.js'));

    gulp.src('target/dist/img/*.png')
        .pipe(ex.real('target/dist/img/*.png'));
    gulp.src('target/dist/img/*.ico')
        .pipe(ex.real('target/dist/img/*.ico'));
    gulp.src('target/dist/fonts/*')
        .pipe(ex.real('target/dist/fonts/glyphicons-halflings-regular.*'));
});

gulp.task('create-package', ['dist'], function () {
    gulp.src(['tasks/*.js', '!tasks/karma.conf.js'])
        .pipe(gulp.dest('target/gript/tasks'));
    gulp.src('package/sample_configs/**/*')
        .pipe(gulp.dest('target/gript/sample_configs'));
    gulp.src('package/sample_configs/.*')
        .pipe(gulp.dest('target/gript/sample_configs'));
    gulp.src('package/img/*.*')
        .pipe(gulp.dest('target/gript/img'));
    gulp.src(['package/*', '!package/karma.conf.js'])
        .pipe(gulp.dest('target/gript'));
    gulp.src('package/karma.conf.js')
        .pipe(gulp.dest('target/gript/tasks'));
    gulp.src('README.md')
        .pipe(gulp.dest('target/gript'));
    gulp.src('CHANGES.md')
        .pipe(gulp.dest('target/gript'));
});

// verify that the npm package contains the files it should
gulp.task('verify-package', function () {
    gulp.src('target/gript/index.js')
        .pipe(ex.real('target/gript/index.js'));
    gulp.src('target/gript/package.json')
        .pipe(ex.real('target/gript/package.json'));
    gulp.src('target/gript/sample_configs/gulpfile.js')
        .pipe(ex.real('target/gript/sample_configs/gulpfile.js'));
    gulp.src('target/gript/sample_configs/.eslintrc.yml')
        .pipe(ex.real('target/gript/sample_configs/.eslintrc.yml'));
    gulp.src('target/gript/sample_configs/.scss-lint.yml')
        .pipe(ex.real('target/gript/sample_configs/.scss-lint.yml'));
    gulp.src('target/gript/sample_configs/bower.json')
        .pipe(ex.real('target/gript/sample_configs/bower.json'));
    gulp.src('target/gript/tasks/*.gulp.js')
        .pipe(ex.real('target/gript/tasks/*.gulp.js'));
    gulp.src('target/gript/tasks/*.conf.js')
        .pipe(ex.real('target/gript/tasks/karma.conf.js'));
    gulp.src('target/gript/README.md')
        .pipe(ex.real('target/gript/README.md'));
    gulp.src('target/gript/CHANGES.md')
        .pipe(ex.real('target/gript/CHANGES.md'));
});
